### PR TITLE
Describe efficiently creating a Vec of zeroes

### PR DESCRIPTION
**Pull request by Shnatsel**

### CODE REVIEW

**Comment by nnethercote** [comment]
I wonder if this needs to be said. How else would you intialize a `Vec` of zeroes?

**Comment by Shnatsel** [comment]
I've seen a developer use `.resize(0, length)`, realize it's too slow, change it to `mem::uninitialized`, and end up with a security vulnerability. [I wrote an entire blog post about this.](https://shnatsel.medium.com/how-ive-found-vulnerability-in-a-popular-rust-crate-and-you-can-too-3db081a67fb)

**Comment by Shnatsel** [comment]
The standard library documentation did add a notice about it after my blog post - open the [`Vec` docs](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html) and Ctrl+F "This may be more efficient than performing allocation and initialization in separate steps, especially when initializing a vector of zeros:"

File: src/standard-library-types.md [hunk 1]
```diff
@@ -41,6 +41,13 @@ not preserve ordering, but is O(1).
 equivalent method for other collection types such as `String`, `HashSet`, and
 `HashMap`.
 
+If you need to create a `Vec` filled with zeroes, the fastest way to do so
+is `vec![0; lenght]`(https://doc.rust-lang.org/stable/std/macro.vec.html).
```
Comment (bobrik) [inline src/standard-library-types.md:45]: `length`

**Comment by bobrik** [comment]
The fix for the bug you reported:

* https://github.com/ruuda/claxon/commit/fec24678c7086aa4b2528bd7542f31d978b81b90

It replaces `unsafe` code. Using [std](https://doc.rust-lang.org/std/index.html)::[mem](https://doc.rust-lang.org/std/mem/index.html)::[uninitialized](https://doc.rust-lang.org/std/mem/fn.uninitialized.html#) (deprecated) and new [std](https://doc.rust-lang.org/std/index.html)::[mem](https://doc.rust-lang.org/std/mem/index.html)::[MaybeUninit](https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#) are also `unsafe`.

**Comment by nnethercote** [comment]
This is a good suggestion. I'm picky about the text of the perf book so I ended up rewriting the paragraph from scratch, in https://github.com/nnethercote/perf-book/commit/3e165fbb081e14afe6d50ed2df7e7300e209b552, so there is no need to merge this PR. Thank you for the contribution!

BTW, from your blog post:

> There is a tool that can detect reads from uninitialized memory, called [Memory Sanitizer](https://clang.llvm.org/docs/MemorySanitizer.html), but it currently [doesn’t work](https://github.com/rust-lang/rust/issues/39610) with Rust standard library. So unless you completely avoid using Rust standard library, there is no tool that let you detect reads from uninitialized memory in Rust.

Valgrind's Memcheck tool detects unsafe uses of uninitialized memory. Also, miri might be able to as well, though I'm not sure about that.

**Comment by Shnatsel** [comment]
Since the wording is being changed anyway, should I create issues instead of pull requests in the future?

**Comment by nnethercote** [comment]
Probably makes sense, yes.

**Pull request closed without merging**